addons/base_import/models.py

```diff
@@ -158,16 +160,33 @@ def get_fields(self, cr, uid, model, context=None,
         return fields
 
     def _read_file(self, file_type, record, options):
+        # guess mimetype from file content
+        mimetype = guess_mimetype(record.file)
+        (file_extension, handler, req) = FILE_TYPE_DICT.get(mimetype, (None, None, None))
+        if handler:
+            try: return getattr(self, '_read_' + file_extension)(record, options)
+            except Exception:
+                _logger.warn("Failed to read file '%s' (transient id %d) using guessed mimetype %s",
+                             record.file_name or '<unknown>', record.id, mimetype)
+
+        # try reading with user-provided mimetype
         (file_extension, handler, req) = FILE_TYPE_DICT.get(file_type, (None, None, None))
         if handler:
-            return getattr(self, '_read_' + file_extension)(record, options)
+            try: return getattr(self, '_read_' + file_extension)(record, options)
+            except Exception:
+                _logger.warn("Failed to read file '%s' (transient id %d) using user-provided mimetype %s",
+                             record.file_name or '<unknown>', record.id, file_type)
+
```

Comment:
Note that just this might be sufficient to fix the issue of CSV files sent with Excel installed: this tries reading the file with xlrd, which blows up, which falls back on by-extension guessing.